application/src/main/java/org/opentripplanner/model/impl/SubmodeMappingModule.java

```diff
@@ -0,0 +1,103 @@
+package org.opentripplanner.model.impl;
+
+import com.csvreader.CsvReader;
+import dagger.Module;
+import dagger.Provides;
+import jakarta.inject.Singleton;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import org.opentripplanner.datastore.api.DataSource;
+import org.opentripplanner.framework.application.OtpAppException;
+import org.opentripplanner.graph_builder.GraphBuilderDataSources;
+import org.opentripplanner.graph_builder.model.GraphBuilderModule;
+import org.opentripplanner.routing.graph.Graph;
+import org.opentripplanner.transit.model.basic.TransitMode;
+
+@Module
+public class SubmodeMappingModule implements GraphBuilderModule {
+
+  private static final String INPUT_FEED_TYPE = "Input feed type";
+  private static final String INPUT_LABEL = "Input label";
+  private static final String GTFS_ROUTE_TYPE = "GTFS route type";
+  private static final String NETEX_SUBMODE = "NeTEx submode";
+  private static final String REPLACEMENT_MODE = "Replacement mode";
+  private static final String[] ALL = {
+    INPUT_FEED_TYPE,
+    INPUT_LABEL,
+    GTFS_ROUTE_TYPE,
+    NETEX_SUBMODE,
+    REPLACEMENT_MODE,
+  };
+
+  private final GraphBuilderDataSources graphBuilderDataSources;
+  private final Graph graph;
+
+  public SubmodeMappingModule(GraphBuilderDataSources graphBuilderDataSources, Graph graph) {
+    this.graphBuilderDataSources = graphBuilderDataSources;
+    this.graph = graph;
+  }
+
+  @Provides
+  @Singleton
+  public static SubmodeMappingService submodeMappingService(Graph graph) {
+    return new SubmodeMappingService(graph.submodeMapping);
+  }
+
+  public void read(DataSource dataSource, Map<SubmodeMappingMatcher, SubmodeMappingRow> map) {
```

Comment:
Does this or `useDefaultMapping` need to be public or could they be private methods?